### PR TITLE
Working evaluate function

### LINKED ISSUE #14: Evaluate function that plots a subset of image + caption in inference

### PR DESCRIPTION
**Pull request by amolero6**
close #14

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -29,40 +30,15 @@ def train_single_epoch(epoch, model, train_loader, optimizer, criterion, device,
...
-		print('-'*89)
-		print(f'| epoch {epoch:3d} | {i:5d}/{len(train_loader):5d} batches | loss {total_loss:5.2f} | ppl {math.exp(total_loss):8.2f}')
-		print('-'*89)
-		print(f'Gen: {model.vocab.generate_caption(torch.argmax(output[0].transpose(1,0), dim=-1))}')
-		print(f'Exp: {model.vocab.generate_caption(target[0,1:])}')
-		print('-'*89)
-		candidate_corpus = [model.vocab.generate_caption(torch.argmax(output[0].transpose(1, 0), dim=-1))]
-		reference_corpus = [model.vocab.generate_caption(target[0, 1:])]
-		print(f'Bleu score: {bleu_score(candidate_corpus, reference_corpus)}')
+		if i % 10==0:
```
Comment (lucaswerner90) [inline train.py:33]: could we remove the `total_loss` variable and just use the `loss.item()` ? It's not related with your changes but since you're modifying this already...

File: evaluate.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import torch
+import nltk
+from torch.nn.functional import pad
+from torch.nn.utils.rnn import pad_packed_sequence, pad_sequence
+from nltk.translate.bleu_score import corpus_bleu
+from model.visualization import Visualization
+
+
+def evaluate(model, test_loader, vocab, device, criterion):  # TODO:add device
```
Comment (lucaswerner90) [inline evaluate.py:9]: remove the #TODO comment

Comment (lucaswerner90) [reply evaluate.py:9]: also, we don't use the `criterion` parameter right?

File: model/main.py [hunk 3]
```diff
@@ -46,14 +46,19 @@ def inference(self, image, vocab:Vocabulary):
...
+		sentence = [word.item()]
+		attention_weights=[]
 		for i in range(self.caption_max_length):
 			alphas, weighted_features = self.attention.forward(image_features, hidden)
 			predictions_t, hidden = self.decoder.forward(weighted_features, word, hidden)
 			word = torch.argmax(predictions_t, dim=-1)
-			sentence.append(word.tolist())
-			
-			if word[0].tolist()==vocab.word_to_index['<END>']:
+			if word[0].item()==vocab.word_to_index['<END>']:
```
Comment (lucaswerner90) [inline model/main.py:55]: I see we send the `Vocabulary` on the constructor, so `self.vocab` is available here. Could we just use that one? So we don't send it as a parameter?

File: train.py (hunk 1 above)
Comment (lucaswerner90) [inline train.py:33]: `if i % log_interval == 0` ?

File: train.py [hunk 4]
```diff
@@ -75,22 +51,24 @@ def save_model(model, epoch):
...
 	Executes model training. Saves model to a file every epoch.
 	"""	
 
 	for epoch in range(1,num_epochs+1):
-
-		train_single_epoch(epoch, model, train_loader,optimizer, criterion, device, log_interval)
+		epoch_start_time=time.time()
+		train_single_epoch(epoch=epoch,model=model,train_loader=train_loader,optimizer=optimizer, criterion=criterion,device= device,log_interval=log_interval)
 
 		if epoch % 5 == 0:
```
Comment (lucaswerner90) [inline train.py:63]: we should use the `log_interval` variable here right?

File: train.py [hunk 5]
```diff
@@ -75,22 +51,24 @@ def save_model(model, epoch):
 	}
 	torch.save(model_state, filename)
 
-def train(num_epochs, model, train_loader,test_loader, optimizer, criterion, device,log_interval):
+def train(num_epochs, model, train_loader,test_loader, optimizer, criterion, device,log_interval,vocab):
 	"""
 	Executes model training. Saves model to a file every epoch.
```
Comment (lucaswerner90) [inline train.py:56]: we should modify the description of the function here since we don't "save the model to a file every epoch"

**Review by lucaswerner90 — CHANGES_REQUESTED** [review]

File: evaluate.py (hunk 2 above)
Comment (amolero6) [reply evaluate.py:9]: Yep, taken out

File: train.py (hunk 5 above)
Comment (amolero6) [reply train.py:56]: Done!

File: train.py (hunk 1 above)
Comment (amolero6) [reply train.py:33]: Let's discuss this to have it all the same, there are different intervals in some places

Comment (amolero6) [reply train.py:33]: Let's discuss it on Friday. I propose to set this as a separate issue to also tackle the log interval and the spaces/tab issue

File: train.py (hunk 4 above)
Comment (amolero6) [reply train.py:63]: Same as for the other comment

**Review by lucaswerner90 — APPROVED** [review]
